alf/networks/network.py

```diff
@@ -84,35 +91,121 @@ def _capture_init(self, *args, **kwargs):
 
 @six.add_metaclass(_NetworkMeta)
 class Network(nn.Module):
-    """Base extension to nn.Module to simplify copy operations."""
+    """A base class for various networks.
+
+    Base extension to nn.Module to simplify copy operations.
 
-    def __init__(self, input_tensor_spec, state_spec, name):
-        """Creates an instance of `Network`.
+    It also handles complex nested inputs.
+    """
 
+    def __init__(self,
+                 input_tensor_spec,
+                 input_preprocessors=None,
+                 preprocessing_combiner=None,
+                 skip_input_preprocessing=False,
+                 name="Network"):
+        """
         Args:
-            input_tensor_spec (nested TensorSpec):  Representing the
-                input observations.
-            state_spec (nested TensorSpec): Representing the state needed by
-                the network. Use () if none.
-            name (str): A string representing the name of the network.
+            input_tensor_spec (nested TensorSpec): the (nested) tensor spec of
+                the input. If nested, then `preprocessing_combiner` must not be
+                None.
+            input_preprocessors (nested InputPreprocessor): a nest of
+                `InputPreprocessor`, each of which will be applied to the
+                corresponding input. If not None, then it must
+                have the same structure with `input_tensor_spec` (after reshaping).
+                If any element is None, then it will be treated as alf.layers.identity.
+                This arg is helpful if you want to have separate preprocessings
+                for different inputs by configuring a gin file without changing
+                the code. For example, embedding a discrete input before concatenating
+                it to another continuous vector.
+            preprocessing_combiner (NestCombiner): preprocessing called on
+                complex inputs. Note that this combiner must also accept
+                `input_tensor_spec` as the input to compute the processed
+                tensor spec. For example, see `alf.nest.utils.NestConcat`. This
+                arg is helpful if you want to combine inputs by configuring a
+                gin file without changing the code.
+            skip_input_preprocessing (bool): If True, skip preprocessing and
```

Comment:
Can't this be inferred from `input_preprocessors`? if input_preprocessors is None, then skip.

Reply:
If `input_preprocessors` is None, the user can still use the combiner (e.g., concat)